src/state/slices/opportunitiesSlice/resolvers/foxFarming/contractManager.ts

```diff
@@ -1,24 +1,67 @@
-import type { ContractInterface } from '@ethersproject/contracts'
-import { Contract } from '@ethersproject/contracts'
 import type { Fetcher, Token } from '@uniswap/sdk'
 import type { providers } from 'ethers'
 import memoize from 'lodash/memoize'
+import { FOX_TOKEN_CONTRACT_ADDRESS } from 'plugins/foxPage/const'
 import { getEthersProvider } from 'plugins/foxPage/utils'
 
+import {
+  foxEthLpContractAddress,
+  foxEthStakingContractAddressV1,
+  foxEthStakingContractAddressV2,
+  foxEthStakingContractAddressV3,
+  foxEthStakingContractAddressV4,
+  foxEthStakingContractAddressV5,
+  uniswapV2Router02ContractAddress,
+} from '../../constants'
+import {
+  ERC20ABI__factory,
+  FarmingAbi__factory,
+  IUniswapV2Pair__factory,
+  IUniswapV2Router02__factory,
+} from './contracts/factories'
+
+type KnownContract<T extends KNOWN_CONTRACTS_ADDRESSES> = ReturnType<
+  typeof CONTRACT_ADDRESS_TO_TYPECHAIN_CONTRACT[T]['connect']
+>
+
+type KNOWN_CONTRACTS_ADDRESSES =
+  | typeof foxEthLpContractAddress
+  | typeof foxEthStakingContractAddressV1
+  | typeof foxEthStakingContractAddressV2
+  | typeof foxEthStakingContractAddressV3
+  | typeof foxEthStakingContractAddressV4
+  | typeof foxEthStakingContractAddressV5
+  | typeof FOX_TOKEN_CONTRACT_ADDRESS
+  | typeof uniswapV2Router02ContractAddress
+
 type DefinedContract = {
-  contract: Contract
-  address: string
+  contract: KnownContract<KNOWN_CONTRACTS_ADDRESSES>
+  address: KNOWN_CONTRACTS_ADDRESSES
 }
 
 const definedContracts: DefinedContract[] = []
 
-export const getOrCreateContract = (address: string, abi: ContractInterface): Contract => {
-  const definedContract = definedContracts.find(contract => contract.address === address)
-  if (definedContract && definedContract.contract) return definedContract.contract
+export const CONTRACT_ADDRESS_TO_TYPECHAIN_CONTRACT = {
+  [foxEthLpContractAddress]: IUniswapV2Pair__factory,
+  [foxEthStakingContractAddressV1]: FarmingAbi__factory,
+  [foxEthStakingContractAddressV2]: FarmingAbi__factory,
+  [foxEthStakingContractAddressV3]: FarmingAbi__factory,
+  [foxEthStakingContractAddressV4]: FarmingAbi__factory,
+  [foxEthStakingContractAddressV5]: FarmingAbi__factory,
+  [FOX_TOKEN_CONTRACT_ADDRESS]: ERC20ABI__factory,
+  [uniswapV2Router02ContractAddress]: IUniswapV2Router02__factory,
+} as const
 
-  const contract = new Contract(address, abi, ethersProvider)
+export const getOrCreateContract = <T extends KNOWN_CONTRACTS_ADDRESSES>(
+  address: T,
```

Comment:
The second guts of this PR - since addresses are literals, we are able to get a factory from it. And since factories are type-safe, we are able to get a type-safe contract as a return type, as the return value of calling `connect` on a factory